QModManager/API/SMLHelper/Patchers/TechTypePatcher.cs

```diff
@@ -0,0 +1,180 @@
+
+                if (bannedIndices.Contains(currentTechTypeKey))
+                    continue; // Already exists in list
+
+                bannedIndices.Add(currentTechTypeKey);
+            }
+
+            if (bannedIndices.Count > 0)
+                Logger.Debug($"Finished known TechTypes exclusion. {bannedIndices.Count} IDs were added in ban list.");
+
+            return bannedIndices;
+        }
+
+        #region Patches
+
+        internal static void Patch(HarmonyInstance harmony)
+        {
+            harmony.Patch(AccessTools.Method(typeof(Enum), "GetValues"),
+                postfix: new HarmonyMethod(AccessTools.Method(typeof(TechTypePatcher), "Postfix_GetValues")));
+
+            harmony.Patch(AccessTools.Method(typeof(Enum), "IsDefined"),
+                prefix: new HarmonyMethod(AccessTools.Method(typeof(TechTypePatcher), "Prefix_IsDefined")));
+
+            harmony.Patch(AccessTools.Method(typeof(Enum), "Parse", new Type[] { typeof(Type), typeof(string), typeof(bool) }),
+                prefix: new HarmonyMethod(AccessTools.Method(typeof(TechTypePatcher), "Prefix_Parse")));
+
+            harmony.Patch(AccessTools.Method(typeof(TechType), "ToString", new Type[] { }),
+                prefix: new HarmonyMethod(AccessTools.Method(typeof(TechTypePatcher), "Prefix_ToString")));
+
+            Logger.Info($"Added {cacheManager.ModdedKeys.Count()} TechTypes succesfully into the game.");
+
+            Logger.Debug("TechTypePatcher is done.");
+        }
+
+        private static void Postfix_GetValues(Type enumType, ref Array __result)
+        {
+            if (enumType.Equals(typeof(TechType)))
+            {
+                var listArray = new List<TechType>();
+                foreach (object obj in __result)
+                {
+                    listArray.Add((TechType)obj);
+                }
+
+                listArray.AddRange(cacheManager.ModdedKeys);
+
+                __result = listArray.ToArray();
+            }
+        }
+
+        private static bool Prefix_IsDefined(Type enumType, object value, ref bool __result)
+        {
+            if (TooltipPatcher.DisableEnumIsDefinedPatch) return true;
+
+            if (enumType.Equals(typeof(TechType)))
+            {
+                if (cacheManager.ContainsKey((TechType)value))
+                {
+                    __result = true;
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool Prefix_Parse(Type enumType, string value, bool ignoreCase, ref object __result)
+        {
+            if (enumType.Equals(typeof(TechType)))
+            {
+                if (cacheManager.TryParse(value, out TechType techType))
+                {
+                    __result = techType;
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool Prefix_ToString(Enum __instance, ref string __result)
```

Comment:
This could also patch non-custom TechTypes to call .AsString on them

Reply:
Hold off on this idea for now. There could be potential here for both useful things and also outright breaking things.  
Let's return to this one after we've sorted out some of the other topics we're trying to address as part of this merge.